### PR TITLE
Rebuild Android apk when Dart source is modified

### LINKED ISSUE #7014: 'flutter run' doesn't rebuild the app if it's changed
cc @johnmccutchan

### PR DESCRIPTION
**Pull request by johnmccutchan**
- [x] Wire up dependency checker and plumb flags down to the right place.
- [x] Refactor some more common bits between the hot and non-hot runners.

Fixes #7014

### CODE REVIEW

**Comment by johnmccutchan** [comment]
@devoncarew

File: packages/flutter_tools/lib/src/application_package.dart [hunk 1]
```diff
@@ -28,6 +28,8 @@ abstract class ApplicationPackage {
 
   String get displayName => name;
 
+  String get packagePath => null;
```
Comment (devoncarew) [inline packages/flutter_tools/lib/src/application_package.dart]: The class is abstract - do we need to provide an implementation here (`=> null`) - do all subclasses override this?

Comment (devoncarew) [inline packages/flutter_tools/lib/src/application_package.dart]: Ah, only overridden for APKs. Perhaps make this method abstract and have all clients implement (even with `=> null` implementations)?

**Comment by devoncarew** [comment]
lgtm!

**Comment by eturk1** [comment]
Issue #7014 was closed but it's still an issue. 
Changing dart code doesn't get the APK rebuilt & debug code windows shows old code before edits.

I can manually delete build\app\intermediates dir to get the debug windows to be the current code BUT...

**Pull request merged**